File: usage-daemon/src/providers/github.js
// GitHub usage provider plugin — descriptor interface.
//
// GitHub's REST API rate-limit status, free and stable:
//   GET https://api.github.com/rate_limit
//     Authorization: Bearer <PAT>
//     -> { resources: { core: {limit,remaining,reset,used}, search: {...}, graphql: {...}, ... } }
// Confirmed live (hit unauthenticated during research, got the real current
// shape back). This measures REST/GraphQL/search API CALL BUDGET, not
// Copilot or Actions consumption — those live behind separate billing
// endpoints (GET /users/{username}/settings/billing/premium_request/usage
// etc, need `Plan: read` on a fine-grained PAT) that this plugin does NOT
// cover yet — two open questions there (does grossQuantity include
// under-allowance usage, is the enhanced billing platform rolled out to this
// account) need resolving against a real token before building that part.
//
// parse() is a PURE function of the raw JSON text.

export const CORE_COLOR = '#0072B2';   // Okabe-Ito blue
export const SEARCH_COLOR = '#E69F00'; // Okabe-Ito orange
export const GRAPHQL_COLOR = '#CC79A7'; // Okabe-Ito reddish-purple

export const ID = 'github';
export const LABEL = 'GitHub';

const API_URL = 'https://api.github.com';
const RATE_LIMIT_PATH = '/rate_limit';
const USER_AGENT = 'usage-daemon/0.1';
const API_VERSION = '2022-11-28';

export class AuthExpiredError extends Error {
  constructor(msg = 'GitHub API token missing or invalid') {
    super(msg);
    this.code = 'auth_expired';
  }
}

export class RateLimitedError extends Error {
  constructor(retryAfter = null) {
    super('rate_limited');
    this.code = 'rate_limited';
    this.retryAfter = retryAfter;
  }
}

function windowFor(id, label, letter, color, resource) {
  if (!resource || typeof resource.limit !== 'number' || typeof resource.remaining !== 'number') return null;
  return {
    id,
    label,
    letter,
    pct: resource.limit > 0 ? Math.max(0, Math.min(100, (100 * (resource.limit - resource.remaining)) / resource.limit)) : null,
    used: resource.remaining, // remaining calls, not consumed
    used_is_remaining: true,
    cap: resource.limit,
    unit: 'calls',
    resets_at: typeof resource.reset === 'number' ? new Date(resource.reset * 1000).toISOString() : null,
    color,
    will_deplete: false,
  };
}

// Pure function of the raw JSON text — no fs/network.
export function parse(raw) {
  let env;
  try {
    env = typeof raw === 'string' ? JSON.parse(raw) : raw;
  } catch {
    throw new AuthExpiredError('unparseable GitHub rate_limit response');
  }
  if (!env || typeof env !== 'object' || !env.resources) {
    throw new AuthExpiredError('no usable GitHub rate_limit figures');
  }

  const windows = [
    windowFor('core', 'REST Calls', 'Rc', CORE_COLOR, env.resources.core),
    windowFor('search', 'Search Calls', 'Sr', SEARCH_COLOR, env.resources.search),
    windowFor('graphql', 'GraphQL Calls', 'Gq', GRAPHQL_COLOR, env.resources.graphql),
  ].filter(Boolean);

  if (windows.length === 0) {
    throw new AuthExpiredError('no usable GitHub rate_limit figures');
  }

  return {
    tier: null,
    windows,
    segments: [],
    // carried out of parse so meta() can surface the raw figures too.
    _github: {
      core_remaining: env.resources.core?.remaining ?? null,
      core_limit: env.resources.core?.limit ?? null,
    },
  };
}

function createProvider() {
  let apiKey = null;
  let lastStats = null; // surfaced via meta()

  return {
    id: ID,
    label: LABEL,
    auth: { kind: 'token' },

    config() {
      return {
        id: ID,
        label: LABEL,
        usageUrl: 'https://github.com/settings/tokens',
        auth: { kind: 'token' },
        category: 'support',
        windows: [{ id: 'core', label: 'REST Calls', color: CORE_COLOR }],
        tiers: [],
      };
    },

    configure(cfg = {}) {
      // !== undefined so configure({api_key:''}) can explicitly clear it.
      if (cfg.api_key !== undefined) apiKey = cfg.api_key ? String(cfg.api_key).trim() : null;
    },

    async setAuth(payload) {
      apiKey = String(payload ?? '').trim() || null;
    },

    async fetch() {
      if (!apiKey) throw new AuthExpiredError('no GitHub API token configured');

      const res = await fetch(`${API_URL}${RATE_LIMIT_PATH}`, {
        headers: {
          Authorization: `Bearer ${apiKey}`,
          'User-Agent': USER_AGENT,
          Accept: 'application/vnd.github+json',
          'X-GitHub-Api-Version': API_VERSION,
        },
      });

      if (res.status === 401 || res.status === 403) throw new AuthExpiredError();
      if (res.status === 429) {
        throw new RateLimitedError(Number(res.headers.get('retry-after')) || null);
      }
      if (!res.ok) throw new Error(`api.github.com HTTP ${res.status}`);

      const body = await res.json();
      lastStats = parse(JSON.stringify(body))._github;
      return JSON.stringify(body);
    },

    intervalSeconds() {
      return 300;
    },

    meta() {
      return lastStats
        ? {
            core_remaining: lastStats.core_remaining,
            core_limit: lastStats.core_limit,
          }
        : {};
    },

    parse,
  };
}

export { createProvider };

File: usage-daemon-v3/usage_daemon/providers/github.py
"""GitHub usage provider plugin (port of src/providers/github.js).

REST/GraphQL/search API call budget from GET /rate_limit with a PAT. Measures
API call budget only — not Copilot/Actions consumption (separate billing
endpoints, not covered yet). used carries remaining calls with
used_is_remaining: true, mirroring the JS.
"""

from __future__ import annotations

import datetime as _dt
import json

from ..errors import AuthExpiredError, RateLimitedError
from ..httputil import create_client

CORE_COLOR = "#0072B2"  # Okabe-Ito blue
SEARCH_COLOR = "#E69F00"  # Okabe-Ito orange
GRAPHQL_COLOR = "#CC79A7"  # Okabe-Ito reddish-purple

ID = "github"
LABEL = "GitHub"

API_URL = "https://api.github.com"
RATE_LIMIT_PATH = "/rate_limit"
USER_AGENT = "usage-daemon/0.1"
API_VERSION = "2022-11-28"


def _num(v) -> bool:
    return isinstance(v, (int, float)) and not isinstance(v, bool)


def _window_for(id, label, letter, color, resource) -> dict | None:
    if not resource or not _num(resource.get("limit")) or not _num(resource.get("remaining")):
        return None
    return {
        "id": id,
        "label": label,
        "letter": letter,
        "pct": max(0, min(100, (100 * (resource["limit"] - resource["remaining"])) / resource["limit"]))
        if resource["limit"] > 0
        else None,
        "used": resource["remaining"],  # remaining calls, not consumed
        "used_is_remaining": True,
        "cap": resource["limit"],
        "unit": "calls",
        "resets_at": _dt.datetime.fromtimestamp(resource["reset"], _dt.timezone.utc).isoformat()
        if _num(resource.get("reset"))
        else None,
        "color": color,
        "will_deplete": False,
    }


def _load(raw):
    try:
        env = json.loads(raw) if isinstance(raw, str) else raw
    except Exception:
        raise AuthExpiredError("unparseable GitHub rate_limit response")
    if not isinstance(env, dict):
        raise AuthExpiredError("unparseable GitHub rate_limit response")
    return env


def parse(raw) -> dict:
    """Pure function of the raw JSON text — no fs/network."""
    env = _load(raw)
    if not env.get("resources"):
        raise AuthExpiredError("no usable GitHub rate_limit figures")

    windows = [
        w
        for w in (
            _window_for("core", "REST Calls", "Rc", CORE_COLOR, env["resources"].get("core")),
            _window_for("search", "Search Calls", "Sr", SEARCH_COLOR, env["resources"].get("search")),
            _window_for("graphql", "GraphQL Calls", "Gq", GRAPHQL_COLOR, env["resources"].get("graphql")),
        )
        if w is not None
    ]
    if not windows:
        raise AuthExpiredError("no usable GitHub rate_limit figures")

    core = env["resources"].get("core") or {}
    return {
        "tier": None,
        "windows": windows,
        "segments": [],
        # carried out of parse so meta() can surface the raw figures too.
        "_github": {
            "core_remaining": core.get("remaining") if _num(core.get("remaining")) else None,
            "core_limit": core.get("limit") if _num(core.get("limit")) else None,
        },
    }


def create() -> dict:
    state = {"api_key": None, "last_stats": None}

    def config() -> dict:
        return {
            "id": ID,
            "label": LABEL,
            "usageUrl": "https://github.com/settings/tokens",
            "auth": {"kind": "token"},
            "category": "support",
            "windows": [{"id": "core", "label": "REST Calls", "color": CORE_COLOR}],
            "tiers": [],
        }

    def configure(cfg: dict | None = None) -> None:
        cfg = cfg or {}
        if "api_key" in cfg:
            state["api_key"] = str(cfg["api_key"]).strip() if cfg["api_key"] else None

    async def set_auth(payload: str) -> None:
        state["api_key"] = str(payload or "").strip() or None

    async def fetch() -> str:
        if not state["api_key"]:
            raise AuthExpiredError("no GitHub API token configured")

        c = create_client()
        try:
            res = await c.get(
                f"{API_URL}{RATE_LIMIT_PATH}",
                headers={
                    "Authorization": f"Bearer {state['api_key']}",
                    "User-Agent": USER_AGENT,
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": API_VERSION,
                },
            )
        finally:
            await c.aclose()

        if res.status_code in (401, 403):
            raise AuthExpiredError()
        if res.status_code == 429:
            ra = res.headers.get("retry-after")
            raise RateLimitedError(int(ra) if ra and ra.isdigit() else None)
        if res.status_code >= 400:
            raise RuntimeError(f"api.github.com HTTP {res.status_code}")

        body = res.json()
        state["last_stats"] = parse(json.dumps(body))["_github"]
        return json.dumps(body)

    def interval_seconds() -> int:
        return 300

    def meta() -> dict:
        s = state["last_stats"]
        if not s:
            return {}
        return {"core_remaining": s["core_remaining"], "core_limit": s["core_limit"]}

    return {
        "id": ID,
        "label": LABEL,
        "auth": {"kind": "token"},
        "config": config,
        "configure": configure,
        "set_auth": set_auth,
        "fetch": fetch,
        "interval_seconds": interval_seconds,
        "meta": meta,
        "parse": parse,
    }
